### DISCUSSION TITLE
button message not work on mobile android and ios

### DISCUSSION THREAD

**Discussion by joeengressia** [tulir/whatsmeow#279] [WhatsApp protocol Q&A]
after update library to go.mau.fi/whatsmeow v0.0.0-20221113135643-05e0c80bc669
button message is not work on mobile, but work on whatsapp desktop and whatsapp web.

this is my code

import (
	proto "go.mau.fi/whatsmeow/binary/proto"
	goProto "google.golang.org/protobuf/proto"
)

var btnOptions []*proto.ButtonsMessage_Button
	btnOptions = make([]*proto.ButtonsMessage_Button, len(l.ButtonMessage.ButtonOptions))
	for _, rowBtn := range l.ButtonMessage.ButtonOptions {
		var protoBtn *proto.ButtonsMessage_Button
		protoBtn = &proto.ButtonsMessage_Button{
			ButtonId: goProto.String(rowBtn.ButtonId),
			ButtonText: &proto.ButtonsMessage_Button_ButtonText{
				DisplayText: goProto.String(rowBtn.ButtonText),
			},
			Type: proto.ButtonsMessage_Button_RESPONSE.Enum(),
		}
		btnOptions = append(btnOptions, protoBtn)
	}
msg := &proto.Message{
		ViewOnceMessage: &proto.FutureProofMessage{
			Message: &proto.Message{
				ButtonsMessage: &proto.ButtonsMessage{
					HeaderType: proto.ButtonsMessage_EMPTY.Enum(),
					ContentText: goProto.String(l.ButtonMessage.HeaderText),
					Buttons:     btnOptions,
					FooterText:  goProto.String(l.ButtonMessage.FooterText),
				}, MessageContextInfo: &proto.MessageContextInfo{
					DeviceListMetadataVersion: goProto.Int32(2),
					DeviceListMetadata:        &proto.DeviceListMetadata{},
				}}}}

**Comment by joeengressia** [comment]
solved

**Reply by emacielxp** [reply]
@joeengressia how did you solved this?

**Reply by emacielxp** [reply]
wrapping message with ViewOnceMessage (&waProto.FutureProofMessage) and inside this the ButtonsMessage, made this works for me

**Reply by joeengressia** [reply]
yes like that and make sure the button id and button text are sent according to Facebook's terms. Previously there was an excess of characters sent, so the button did not appear.

![photo_2022-11-24 08 38 34](https://user-images.githubusercontent.com/1888121/203675121-7659e82e-8e7a-4855-97a1-1ae3f1a11f36.jpeg)

**Reply by ahmedRSA** [reply]
> yes like that and make sure the button id and button text are sent according to Facebook's terms. Previously there was an excess of characters sent, so the button did not appear.
> 
> ![photo_2022-11-24 08 38 34](https://user-images.githubusercontent.com/1888121/203675121-7659e82e-8e7a-4855-97a1-1ae3f1a11f36.jpeg)

what is facebook term?

**Reply by jraylan** [reply]
>
> what is facebook term?

<img width="100%" alt="Captura de Tela 2022-11-30 às 18 50 00" src="https://user-images.githubusercontent.com/23512861/204915536-4f8c8a82-52ae-4cd6-ba86-f1fb76244944.png">

[https://developers.facebook.com/docs/whatsapp/on-premises/reference/messages#buttons](https://developers.facebook.com/docs/whatsapp/on-premises/reference/messages#buttons)
